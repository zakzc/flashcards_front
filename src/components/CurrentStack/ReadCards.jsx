import React, { Component } from "react";
// hooks
//import { useDB_Connection } from "../../DB/DB-hook/connection-hook";

export default class ReadCards extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentUser: props.currentUser,
      currentStack: props.currentStack,
    };
  }

  showCurrentStack() {
    if (this.state.currentStack.stackName) {
      return (
        <div id="cardListContainer">
          {this.state.currentStack.cards.map((m, j) => {
            return (
              <div className="flipCard" key={j}>
                <li className="flipCardInner" value={m.front}>
                  <div className="flipCardFront">
                    <div className="cardTextPositioning">{m.front}</div>
                  </div>
                  <div className="flipCardBack">
                    <div className="cardTextPositioning">{m.back}</div>
                  </div>
                  {console.log("card: ", m.front, m.back)}
                </li>
              </div>
            );
          })}
          <div></div>
        </div>
      );
    } else {
      return (
        <div>
          <h3>
            There was an error retrieving your cards. Please, Reload the
            application and if the problem persists, inform the developer.
          </h3>
        </div>
      );
    }
  }

  render() {
    return <React.Fragment>{this.showCurrentStack()}</React.Fragment>;
  }
}
