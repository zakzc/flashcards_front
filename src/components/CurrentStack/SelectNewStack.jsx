import React, { Component } from "react";
// import history from "../../utils/history";
import { Redirect } from "react-router-dom";
// hooks
//import { useDB_Connection } from "../../DB/DB-hook/connection-hook";

export default class SelectNewStack extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userIsLoggedIn: props.userIsLoggedIn,
      currentStack: props.currentStack,
      currentUser: props.currentUser,
      chosenStackId: "",
      chosenStackName: "",
      confirmation: false,
      redirect: false,
    };
    // methods from parent comp
    this.setCurrentStack = props.setCurrentStack;
    //this.selectStackUpdatedStack = props.selectStackUpdatedStack;
    // local methods
    this.stackSelector = this.stackSelector.bind(this);
    this.chooseNewStack = this.chooseNewStack.bind(this);

    this.getNewStackName = this.getNewStackName.bind(this);
    this.confirmNewStack = this.confirmNewStack.bind(this);
    this.renderRedirect = this.renderRedirect.bind(this);
    //this.reloadApp = this.reloadApp.bind(this);
    // views
    this.chooseNewStackView = this.chooseNewStackView.bind(this);
    this.preSelectView = this.preSelectView.bind(this);
    this.afterSelectView = this.afterSelectView.bind(this);
  }

  getNewStackName(newStackID) {
    let listOfStacks = this.state.currentUser.userStacks;
    for (let l in listOfStacks) {
      // console.log("Checking: ", l, listOfStacks[l].stack_id);
      if (newStackID === listOfStacks[l].stack_id) {
        return listOfStacks[l].stack_name;
      }
    }
  }

  // Chooses new option
  chooseNewStack(e) {
    let newStack = e.target.value;
    let newStackName = this.getNewStackName(newStack);
    // console.log("choose stack: ", newStack, "name: ", newStackName);
    this.setState({
      chosenStackId: newStack,
      chosenStackName: newStackName,
      confirmation: true,
    });
  }

  stackSelector() {
    return (
      <select className="select-css" onChange={this.chooseNewStack}>
        <option value="0">Select new stack</option>
        {this.state.currentUser.userStacks.map((s) => {
          return (
            <option key={s.stack_name} value={s.stack_id}>
              {s.stack_name}
            </option>
          );
        })}
      </select>
    );
  }

  renderRedirect() {
    if (this.state.redirect) {
      return (
        <Redirect
          to="/consolidateChanges"
          userIsLoggedIn={this.state.userIsLoggedIn}
        />
      );
    }
  }

  confirmNewStack() {
    this.setState({ redirect: true });
    this.setCurrentStack(this.state.chosenStackId);
  }

  ////////////////
  // View methods
  ////////////////

  confirmationButton() {
    return (
      <div>
        <h3> You chose: {this.state.chosenStackName}</h3>
        {this.renderRedirect()}
        <button
          className="buttonStyle"
          type="button"
          onClick={this.confirmNewStack}
        >
          Confirm change
        </button>
      </div>
    );
  }

  preSelectView() {
    return (
      <div id="StackSelection">
        <h2>
          Hi there. Your current stack of cards is:{" "}
          <strong className="currentSelectedStack">
            {this.state.currentStack.stackName}
            {". "}
          </strong>
        </h2>
        <h3>Which stack do you want to select?</h3>
        {this.stackSelector()}
      </div>
    );
  }

  afterSelectView() {
    return (
      <div id="confirmation">
        {this.state.confirmation ? this.confirmationButton() : <br />}
      </div>
    );
  }

  chooseNewStackView() {
    return (
      <div>
        {this.state.confirmation
          ? this.afterSelectView()
          : this.preSelectView()}
      </div>
    );
  }

  selectNewStackView() {
    return (
      <div id="cardSetView">
        {this.state.userIsLoggedIn ? (
          <div id="loggedInView">{this.chooseNewStackView()}</div>
        ) : (
          <Redirect to="/" />
        )}
      </div>
    );
  }

  render() {
    return <React.Fragment>{this.selectNewStackView()}</React.Fragment>;
  }
}
