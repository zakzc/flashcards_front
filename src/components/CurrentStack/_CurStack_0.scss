#currentStackLayout {
  display: grid;
  grid-template-rows: 1fr auto;
  gap: 20px;
  margin-left: 2%;
}

#saluteTheUser {
  grid-area: salute;
}

#buttonArea {
  grid-area: buttonArea;
  background-color: transparent;
}

#tinyNavBar {
  margin-top: 20px;
}

// Adjusts for responsiveness
@media only screen and (max-width: 767px) {
  #tinyNavBar {
    display: grid;
    gap: 4%;
    grid-template-rows: repeat(5, auto);
    // justify-content: space-around;
    // align-content: center;
    grid-template-areas:
      "."
      "salute"
      "."
      "buttonArea"
      ".";
  }

  #cardListContainer {
    display: inline-block;
    margin-left: -30%;
  }
}

@media only screen and (min-width: 768px) {
  #tinyNavBar {
    display: grid;
    grid-template-columns: repeat(5, auto);
    justify-content: space-around;
    align-content: center;
    grid-template-areas: " . salute . buttonArea .";
  }
}

.cardViewWrap {
  display: grid;
  grid-template-rows: auto;
  justify-content: center;
  align-content: center;
}
