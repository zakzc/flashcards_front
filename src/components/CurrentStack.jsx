import React, { Component } from "react";
import { Redirect } from "react-router-dom";
// internal componentes
import ReadCards from "./CurrentStack/ReadCards";
import SelectNewStack from "./CurrentStack/SelectNewStack";
// Standard Log out page
import UserIsLoggedOut from "./views/UserIsLoggedOut";
// Style
import "../../src/index.css";
// Icons
import Collection from "./icons/collection";
import CheckEye from "./icons/checkEye";

export default class currentStack extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentUser: props.currentUser,
      userIsLoggedIn: props.userIsLoggedIn,
      userStacks: props.userStacks,
      currentStack: props.currentStack,
      readOrSelect: true,
      update: false,
    };
    // components
    this.UserIsLoggedOut = UserIsLoggedOut;
    this.ReadCards = ReadCards;
    this.selectNewStack = SelectNewStack;
    // methods from app.js
    this.setCurrentStack = props.setCurrentStack;
    // Methods
    this.changeStackButton = this.changeStackButton.bind(this);
    this.currentStackLayout = this.currentStackLayout.bind(this);
    this.readSelectSwitch = this.readSelectSwitch.bind(this);
    this.SelectDifferentStackButton = this.SelectDifferentStackButton.bind(
      this
    );
    // Views
    this.currentStackView = this.currentStackView.bind(this);
  }

  readSelectSwitch() {
    console.log("read, select");
    this.setState((state) => ({
      readOrSelect: !state.readOrSelect,
    }));
  }

  SelectDifferentStackButton() {
    return (
      <div id="readSelectButton_Positioning">
        <div className="flipContainer">
          <div className="flipInner">
            <button
              className="buttonStyle"
              type="button"
              onClick={this.readSelectSwitch}
            >
              <div className="flipFront">
                <Collection className="readSelectButtonStyle" />
              </div>
              <div className="flipBack">
                <span className="buttonMessage">Select a different stack</span>
              </div>
            </button>
          </div>
        </div>
      </div>
    );
  }

  ReadStackButton() {
    return (
      <div id="readSelectButton_Positioning">
        <div className="flipContainer">
          <div className="flipInner">
            <button
              className="buttonStyle"
              type="button"
              onClick={this.readSelectSwitch}
            >
              <div className="flipFront">
                <CheckEye className="readSelectButtonStyle" />
              </div>
              <div className="flipBack">
                <span className="buttonMessage">Read current stack</span>
              </div>
            </button>
          </div>
        </div>
      </div>
    );
  }

  changeStackButton() {
    return (
      <div id="buttonArea">
        {this.state.readOrSelect ? (
          <div>{this.SelectDifferentStackButton()}</div>
        ) : (
          <div>{this.ReadStackButton()}</div>
        )}
      </div>
    );
  }

  currentStackLayout() {
    return (
      <React.Fragment>
        <div id="currentStackLayout">
          {this.changeStackButton()}
          <div id="saluteTheUser">
            <h3>Hi {this.state.currentUser.firstName}, welcome back </h3>
            <br />
            <p>The current stack is: {this.state.currentStack.stackName}</p>
          </div>
          <div className="cardViewWrap">
            {this.state.readOrSelect ? (
              <ReadCards
                userIsLoggedIn={this.state.userIsLoggedIn}
                currentUser={this.state.currentUser}
                currentStack={this.state.currentStack}
              />
            ) : (
              <SelectNewStack
                userIsLoggedIn={this.state.userIsLoggedIn}
                currentStack={this.state.currentStack}
                currentUser={this.state.currentUser}
                setCurrentStack={this.setCurrentStack}
              />
            )}
          </div>
        </div>
      </React.Fragment>
    );
  }

  currentStackView() {
    return (
      <div id="currentStackView">
        {this.state.userIsLoggedIn ? (
          <div id="loggedInView">{this.currentStackLayout()}</div>
        ) : (
          <div>
            <Redirect to="/" />
          </div>
        )}
        {/* {this.checkForConsistency()} */}
      </div>
    );
  }

  render() {
    return <React.Fragment>{this.currentStackView()}</React.Fragment>;
  }
}
