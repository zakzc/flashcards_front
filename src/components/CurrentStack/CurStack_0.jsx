import React, { Component } from "react";
import { Redirect } from "react-router-dom";
// internal components
import ReadCards from "./CurStack_ReadCards";
import SelectNewStack from "./CurStack_SelectNewStack";
// Standard Log out page
import UserIsLoggedOut from "../views/UserIsLoggedOut";
// Style
import "../../../src/index.css";
// Icons
import Collection from "../../assets/icons/collection";
import CheckEye from "../../assets/icons/checkEye";

export default class CurrentStack extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentUser: props.currentUser,
      userIsLoggedIn: props.userIsLoggedIn,
      userStacks: props.userStacks,
      currentStack: props.currentStack,
      readOrSelect: true,
      update: false,
    };
    // components
    this.UserIsLoggedOut = UserIsLoggedOut;
    this.ReadCards = ReadCards;
    this.selectNewStack = SelectNewStack;
    // methods from app.js
    this.setCurrentStack = props.setCurrentStack;
    // Methods
    this.changeStackButton = this.changeStackButton.bind(this);
    this.currentStackLayout = this.currentStackLayout.bind(this);
    this.readSelectSwitch = this.readSelectSwitch.bind(this);
    this.SelectDifferentStackButton = this.SelectDifferentStackButton.bind(
      this
    );
    this.currentStackView = this.currentStackView.bind(this);
  }

  ////////////////
  // Implementation methods
  ////////////////

  readSelectSwitch() {
    this.setState((state) => ({
      readOrSelect: !state.readOrSelect,
    }));
  }

  ////////////////
  // View methods
  ////////////////

  SelectDifferentStackButton() {
    return (
      <button
        className="buttonStyle"
        type="button"
        onClick={this.readSelectSwitch}
      >
        <div className="flipContainer">
          <div className="flipInner">
            <div className="flipFront">
              <Collection />
            </div>
            <div className="flipBack">
              <span className="buttonMessage">Select stack</span>
            </div>
          </div>
        </div>
      </button>
    );
  }

  ReadStackButton() {
    return (
      <button
        className="buttonStyle"
        type="button"
        onClick={this.readSelectSwitch}
      >
        <div className="flipContainer">
          <div className="flipInner">
            <div className="flipFront">
              <CheckEye />
            </div>
            <div className="flipBack">
              <span className="buttonMessage">Read stack</span>
            </div>
          </div>
        </div>
      </button>
    );
  }

  changeStackButton() {
    return (
      <div id="buttonArea">
        {this.state.readOrSelect ? (
          <div>{this.SelectDifferentStackButton()}</div>
        ) : (
          <div>{this.ReadStackButton()}</div>
        )}
      </div>
    );
  }

  currentStackLayout() {
    return (
      <React.Fragment>
        <div id="currentStackLayout">
          <div id="tinyNavBar">
            <div id="saluteTheUser">
              <h3>Hi {this.state.currentUser.firstName}, welcome back </h3>
              <br />
              <p>
                The current stack is:{" "}
                <span className="highlight">
                  {this.state.currentStack.stackName}
                </span>
              </p>
              <p>
                Hover over the card to see the back, or click on it if you are
                on
              </p>
              a mobile.
            </div>
            {this.changeStackButton()}
          </div>
          <div className="cardViewWrap">
            {this.state.readOrSelect ? (
              <ReadCards
                userIsLoggedIn={this.state.userIsLoggedIn}
                currentUser={this.state.currentUser}
                currentStack={this.state.currentStack}
              />
            ) : (
              <SelectNewStack
                userIsLoggedIn={this.state.userIsLoggedIn}
                currentStack={this.state.currentStack}
                currentUser={this.state.currentUser}
                setCurrentStack={this.setCurrentStack}
              />
            )}
          </div>
        </div>
      </React.Fragment>
    );
  }

  currentStackView() {
    // console.log("CURRENT STACK");
    return (
      <div id="currentStackView">
        {this.state.userIsLoggedIn ? (
          <div id="loggedInView">{this.currentStackLayout()}</div>
        ) : (
          <div>
            <Redirect to="/" />
          </div>
        )}
        {/* {this.checkForConsistency()} */}
      </div>
    );
  }

  render() {
    return <React.Fragment>{this.currentStackView()}</React.Fragment>;
  }
}
