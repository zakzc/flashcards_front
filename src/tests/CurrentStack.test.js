import React from "react";
import Enzyme from "enzyme";
// import ReactDOM from "react-dom";
// import { render } from "@testing-library/react";
///
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
///
import CurrentStack from "../components/CurrentStack";

Enzyme.configure({ adapter: new Adapter() });

describe("CurrentStack component", () => {
  const currentStack = shallow(<CurrentStack />);

  describe("Initialization process", () => {
    it("Renders the CardSet component", () => {
      expect(currentStack).toMatchSnapshot();
    });

    it("Initializes the State", () => {
      expect(currentStack.state().readOrSelect).toEqual(true);
      expect(currentStack.state().update).toEqual(false);
    });
    describe("Imports", () => {
      test("ReadCards", () => {
        expect(currentStack.find("ReadCards")).toBeTruthy();
      });
      test("SelectNewStack", () => {
        expect(currentStack.find("SelectNewStack")).toBeTruthy();
      });
      test("UserIsLoggedOut", () => {
        expect(currentStack.find("UserIsLoggedOut")).toBeTruthy();
      });
      describe("Svg imports", () => {
        test("Collection", () => {
          expect(currentStack.find("Collection")).toBeTruthy();
        });
        test("CheckEye", () => {
          expect(currentStack.find("CheckEye")).toBeTruthy();
        });
      });
    });
  });
});

// import ReadCards from "./CurrentStack/ReadCards";
// import SelectNewStack from "./CurrentStack/SelectNewStack";
// // Standard Log out page
// import UserIsLoggedOut from "./views/UserIsLoggedOut";
// // Style
// import "../../src/index.css";
// // Icons
// import Collection from "./icons/collection";
// import CheckEye from "./icons/checkEye";
