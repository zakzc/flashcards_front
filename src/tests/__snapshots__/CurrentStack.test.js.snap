// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`
 ___| Select New Stack |___  	Initialization process Renders the ManageCards component 1`] = `ShallowWrapper {}`;

exports[`
 __| Current Stack |__ Initialization process Renders the CardSet component 1`] = `ShallowWrapper {}`;
